File: java/cours/src/session3/TD/MotIterableAvecIterateurEtVisiteur/Mot.java
import java.util.Iterator;

interface Mot extends Iterable<Character> {
	default boolean estMotNonVide(){
		return false;
	}

	default boolean estConcat(){
		return false;
	}

	default char element(){
		throw new UnsupportedOperationException();
	}

	default Mot reste() {
		throw new UnsupportedOperationException();
	}

	default Mot gauche(){
		throw new UnsupportedOperationException();
	}

	default Mot droit(){
		throw new UnsupportedOperationException();
	}

	default Mot vide(){
		return MotVide.VIDE; 
	}

	default Mot motNonVide(char n){
		return new MotNonVide(n, this);
	}

	default Mot concat(Mot ens){
		return new MotConcat(this, ens);
	}

	int taille();

	default boolean estVide(){
		return this.taille() == 0;
	}

	default Iterateur iterateur(){
		return new Iterateur(this);
	}

	default 	
	public Iterator<Character> iterator() {
		return this.iterateur();
	}
	
	default void accept(Visiteur visitor) {
		visitor.visitRec(this);
	}
}

File: typescript/session3/MotIterableAvecIterateur/Mot.ts
import { Char } from "../Char";
import { MotNonVide } from './MotNonVide';
import { MotVide } from './MotVide';
import { MotConcat } from './MotConcat';
import { Iterateur } from "./Iterateur";

export abstract class Mot implements Iterable<Char>{


  public constructor() {

  }
  estMotNonVide(): boolean {
    return false;
  }
  estConcat(): boolean {
    return false;
  }

  element(): Char {
    throw new Error();
  }
  reste(): Mot {
    throw new Error();
  }

  gauche(): Mot {
    throw new Error();
  }

  droit(): Mot {
    throw new Error();
  }

  vide(): Mot {
    return MotVide.VIDE;
  }

  mot(n: Char): Mot {
    return new MotNonVide(n, this);
  }

  concat(ens: Mot): Mot {
    return new MotConcat(this, ens);
  }

  abstract taille(): number;

  estVide(): boolean {
    return this.taille() == 0;
  }

  iterateur(): Iterateur {
    return new Iterateur(this);
  }

  [Symbol.iterator](): Iterator<Char> {
    return this.iterateur();
  }
}